src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -960,5 +1009,48 @@ void ThrowIfNotPeekLockMode()
 
             MessagingEventSource.Log.RegisterOnMessageHandlerStop(this.ClientId);
         }
+
+        public void RegisterPlugin(ServiceBusPlugin serviceBusPlugin)
+        {
+            if (serviceBusPlugin == null)
+            {
+                throw new ArgumentNullException(nameof(serviceBusPlugin), Resources.ArgumentNullOrWhiteSpace.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            if (this.RegisteredPlugins == null)
+            {
+                this.RegisteredPlugins = new List<ServiceBusPlugin>();
+            }
+            else if (this.RegisteredPlugins.Any(p => p.GetType() == serviceBusPlugin.GetType()))
+            {
+                throw new ArgumentException(nameof(serviceBusPlugin), Resources.PluginAlreadyRegistered.FormatForUser(nameof(serviceBusPlugin)));
+            }
+            this.RegisteredPlugins.Add(serviceBusPlugin);
+        }
+
+        public void UnregisterPlugin(ServiceBusPlugin serviceBusPlugin)
```

Comment:
I partly agree with @Sean here. I think we should be able to register plugin only during creation of Queue unless we see a strong reason to change things dynamically.
But at the same time, we might want to have `UnregisterPlugin` or `ClosePlugin` equivalent so that plugin can probably release some resources of its own if required. But this would require us to expose additional methods on plugin. 
So in effect, we might end up having two more functions `OnPluginRegistered` and `OnPluginUnregistered` methods in `ServiceBusPlugin`

---
In reply to: [117356385](https://github.com/Azure/azure-service-bus-dotnet/pull/153#discussion_r117356385) [](ancestors = 117356385)